teuthology/task/__init__.py

```diff
@@ -0,0 +1,116 @@
+        if not isinstance(self.config, dict):
+            raise TypeError("config must be a dict")
+
+    def apply_overrides(self):
+        """
+        Look for an 'overrides' dict in self.ctx.config; look inside that for a
+        dict with the same name as this task. Override any settings in
+        self.config with those overrides
+        """
+        all_overrides = self.ctx.config.get('overrides', dict())
+        if not all_overrides:
+            return
+        task_overrides = all_overrides.get(self.name)
+        if task_overrides:
+            self.log.debug(
+                "Applying overrides for task {name}: {overrides}".format(
+                    name=self.name, overrides=task_overrides)
+            )
+            deep_merge(self.config, task_overrides)
+
+    def find_hosts(self):
+        """
+        Look for a 'hosts' list in self.config. Each item in the list may
+        either be a role or a hostname. The task will then be run against only
+        those hosts which match one (or more) of the roles or hostnames
+        specified.
+        """
+        if not hasattr(self.ctx, 'cluster'):
+            return
+        host_specs = self.config.get('hosts', list())
+        hosts = set()
+        for host_spec in host_specs:
+            role_matches = self.ctx.cluster.only(host_spec)
+            if len(role_matches.remotes) > 0:
+                for remote in role_matches.remotes.keys():
+                    hosts.add(remote)
+            elif isinstance(host_spec, basestring):
+                for remote in self.ctx.cluster.remotes.keys():
+                    if remote.name.split('@')[-1] == host_spec or \
+                            remote.shortname == host_spec:
+                        hosts.add(remote)
+        if not hosts:
+            raise RuntimeError("All target hosts were excluded!")
+        self.hosts = list(hosts)
+        hostnames = [h.shortname for h in self.hosts]
+        self.log.debug("Restricting task {name} to hosts: {hosts}".format(
+            name=self.name, hosts=' '.join(hostnames))
+        )
+        return self.hosts
+
+    def setup(self):
+        """
+        Perform any setup that is needed by the task before it executes
+        """
+        self.apply_overrides()
+        self.find_hosts()
+
+    def begin(self):
+        """
+        Execute the main functionality of the task
+        """
+        pass
+
+    def end(self):
+        """
+        Perform any work needed to stop processes started in begin()
+        """
+        pass
+
+    def teardown(self):
+        """
+        Perform any work needed to restore configuration to a previous state.
+
+        Can be skipped by setting 'skip_teardown' to True in self.config
+        """
+        pass
+
+    def __enter__(self):
+        """
+        When using an instance of the class as a context manager, this method
+        calls self.setup(), then calls self.begin() and returns self.
+        """
+        self.setup()
```

Comment:
as mentioned above. I think that a proxy function of sorts here that calls self.setup() would be better as subclasses will often want to override setup.  I suspect somebody will eventually forget to call super() and we wouldn't apply_overrides or find_hosts.